### PR TITLE
fix: update Nft metadata when toggles are on

### PR DESCRIPTION
**Pull request by sahar-fehri**
## **Description**

Fetches NFT metadata when IPFS toggle or Display media toggle are on

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/23293?quickstart=1)

## **Related issues**

Core PR: https://github.com/MetaMask/core/pull/4008

## **Manual testing steps**

1. Go to settings and disable IPFS toggle and disable display Media toggle
2. Import an NFT (where metadata is hosted on IPFS)
3. Enable IPFS toggle
4. You should see the image on the home page
5. Import an NFT where NFT metadata are hosted on https server
6. Go to settings and enable the display media toggle
7. You should see the NFT metadata on home page

### **Before**

https://github.com/MetaMask/metamask-extension/assets/10994169/c0d09b77-1cd2-495e-aec8-a4ba965ed38f

### **After**

https://github.com/MetaMask/metamask-extension/assets/10994169/06e56b4f-5c1b-4402-b6e2-154daff291d5

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by sahar-fehri** [comment]
update-policies

File: .yarn/patches/@metamask-assets-controllers-patch-74191f09e4.patch [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+         });
+     }
++    /**
++     * Refetches NFT metadata and updates the state.
++     *
++     * @param nfts - Array of nfts
++     * @param networkClientId - The networkClientId that can be used to identify the network client to use for this request.
++     * @param userAddress - The current user address
++     */
++    updateNftMetadata(nfts, networkClientId, userAddress = this.config.selectedAddress) {
```
Comment (sahar-fehri) [inline .yarn/patches/@metamask-assets-controllers-patch-74191f09e4.patch:16]: Here are the steps how this file was generated:
1- Updated node-modules NftController.js (by adding my new fct)
2- `yarn patch @metamask/assets-controllers --update     ` 
3-  I updated the nftController.js file under `/private/var/folders/j2/z5v3y5wd23z_9ldp7tkknszh0000gn/T/xfs-6b6a5943/user `
4- yarn patch-commit -s /private/var/folders/j2/z5v3y5wd23z_9ldp7tkknszh0000gn/T/xfs-6b6a5943/user

File: ui/store/actions.ts [hunk 2]
```diff
@@ -2002,6 +2002,34 @@ export function addNft(
...
+export function updateNftMetadata(
+  nfts: [],
+): ThunkAction<void, MetaMaskReduxState, unknown, AnyAction> {
+  return async (dispatch: MetaMaskReduxDispatch, getState) => {
+    if (nfts.some((nft) => !nft.address)) {
+      throw new Error('MetaMask - Cannot updateNftMetadata without address');
+    }
+    if (nfts.some((nft) => !nft.tokenId)) {
+      throw new Error('MetaMask - Cannot updateNftMetadata without tokenId');
+    }
```
Comment (DDDDDanica) [inline ui/store/actions.ts:2014]:
```suggestion
    if (nfts.some((nft) => !nft.address)) {
      throw new Error('MetaMask - Cannot updateNftMetadata without address');
    }
```

File: ui/components/app/nfts-tab/nfts-tab.js [hunk 3]
```diff
@@ -66,6 +71,37 @@ export default function NftsTab() {
...
+      const allNfts = [...previouslyOwnedCollection.nfts];
+      for (const key in collections) {
+        if (Object.hasOwn(collections, key)) {
+          allNfts.push(...collections[key].nfts);
+        }
+      }
+      const nftsToUpdateMetadata = allNfts.filter(
+        (singleNft) =>
+          !singleNft.name && !singleNft.description && !singleNft.image,
+      );
```
Comment (DDDDDanica) [inline ui/components/app/nfts-tab/nfts-tab.js:89]: Try this to simplify the collecting for `allNfts` if you think it is a good approach :)
```
const allNfts = previouslyOwnedCollection.nfts.concat(
  Object.values(collections).flatMap(collection => collection.nfts)
);
```

**Comment by sahar-fehri** [comment]
closing this in favor of this https://github.com/MetaMask/metamask-extension/pull/23558

**Pull request closed without merging**